pkg/operator/ceph/cluster/mon/spec.go

```diff
@@ -90,22 +103,196 @@ func (c *Cluster) makeDeployment(config *monConfig, hostname string) *extensions
+		// also writes a copy of the config to the default path
+		"--conf=%s", cephconfig.DefaultConfigFilePath(),
+		// It's similarly safe to use the default path for the keyring b/c
+		// 'CreateKeyring' also writes a copy of the keyring to the default path
+		"--keyring=%s", cephconfig.DefaultKeyringFilePath(),
+	}
+
+	logger.Infof("context.NetworkInfo.PublicAddr = %s", c.context.NetworkInfo.PublicAddr)
+	logger.Infof("context.NetworkInfo.ClusterAddr = %s", c.context.NetworkInfo.ClusterAddr)
+
+	// container := c.monContainer(config, c.clusterInfo.FSID)
 	podSpec := v1.PodSpec{
-		Containers:    []v1.Container{container},
+		InitContainers: []v1.Container{
+			{ // Config file init performed by Rook
+				Args: []string{
+					"ceph",
+					"mon",
+					fmt.Sprintf("--config-dir=%s", k8sutil.DataDir),
+					fmt.Sprintf("--name=%s", config.DaemonName),
+					fmt.Sprintf("--port=%d", config.Port),
+					fmt.Sprintf("--fsid=%s", c.clusterInfo.FSID),
+				},
+				Name:            "config-init",
+				Image:           k8sutil.MakeRookImage(c.Version),
+				Env:             c.initEnvVars(config),
+				VolumeMounts:    initMounts(),
+				SecurityContext: podSecurityContext(),
+				// Init container doesn't need to expose ports since it does not communicate
+				Resources: c.resources,
+			},
+			{ // Ceph monmap init performed by mon daemon
+				// Command: []string{"echo"},
+				Command: []string{
+					monmaptoolCommand,
+				},
+				Args: append(
+					[]string{
+						monmapFilePath,
+						"--create",
+						"--clobber",
+						"--fsid", c.clusterInfo.FSID,
+					},
+					monmapAddMonArgs...,
+				),
+				Name: "mon-monmap-init",
+				// TODO; Replace with ceph/ceph:v<vers> image after it's available
+				Image: k8sutil.MakeRookImage(c.Version),
+				// Containers which run daemons do not need env vars
+				VolumeMounts:    daemonMounts(),
+				SecurityContext: podSecurityContext(),
+				Ports:           daemonPorts(config), // TODO: does monmaptool really need ports open?
+				Resources:       c.resources,
+			},
+			{ // Ceph mon filesystem init performed by mon daemon
+				Command: []string{
+					cephMonCommand,
+				},
+				Args: append(
+					[]string{
+						"--mkfs",
+						"--monmap", monmapFilePath,
+					},
+					monCommonArgs...,
+				),
+				Name: "mon-fs-init",
+				// TODO; Replace with ceph/ceph:v<vers> image after it's available
+				Image: k8sutil.MakeRookImage(c.Version),
+				// Containers which run daemons do not need env vars
+				VolumeMounts:    daemonMounts(),
+				SecurityContext: podSecurityContext(),
+				Ports:           daemonPorts(config), // TODO: Does mkfs really need ports open?
+				Resources:       c.resources,
+			},
+		},
+		Containers: []v1.Container{
+			{
+				Command: []string{
+					cephMonCommand,
+				},
+				Args: append(
+					[]string{
+						"--foreground",
+						"--public-addr", joinHostPort(c.context.NetworkInfo.PublicAddr, config.Port),
```

Comment:
@travisn, I'm seeing an issue here that when I get to this line and the one below, the context has empty strings for `NetworkInfo.{Public,Cluster}Addr`. Is this information not generated by the operator yet at this point?

Reply:
That makes sense. So I will get the cluster addr from `config` in the operator, and I will set the the public addr from `PodIPEnvVar`.

Reply:
actually I think you want to switch those two